File: 08-Recursion-Backtracking/cpp/n_queens_ii.cpp
/*
 * Problem: N-Queens II
 * Source:  LeetCode 52 - https://leetcode.com/problems/n-queens-ii/
 * 
 * Approach: Backtracking with bitmask optimization. Track occupied columns,
 *           diagonals, and anti-diagonals using bitmasks for O(1) validity checks.
 * Time Complexity:  O(n!)
 * Space Complexity: O(n)
 * 
 * Tags: recursion, backtracking, bitmask
 */

#include <bits/stdc++.h>
using namespace std;

int totalNQueens(int n) {
    int count = 0;
    
    function<void(int, int, int, int)> solve = [&](int row, int cols, int diag, int antiDiag) {
        if (row == n) {
            count++;
            return;
        }
        
        int available = ((1 << n) - 1) & ~(cols | diag | antiDiag);
        
        while (available) {
            int pos = available & (-available); // lowest set bit
            available ^= pos;
            
            solve(row + 1,
                  cols | pos,
                  (diag | pos) << 1,
                  (antiDiag | pos) >> 1);
        }
    };
    
    solve(0, 0, 0, 0);
    return count;
}

int main() {
    cout << "N=4: " << totalNQueens(4) << endl;  // Output: 2
    cout << "N=8: " << totalNQueens(8) << endl;  // Output: 92
    cout << "N=1: " << totalNQueens(1) << endl;  // Output: 1
    
    return 0;
}

File: 08-Recursion-Backtracking/java/NQueensII.java
/*
 * Problem: N-Queens II
 * Source:  LeetCode 52 - https://leetcode.com/problems/n-queens-ii/
 * 
 * Approach: Backtracking with boolean arrays for columns, diags, anti-diags.
 * Time Complexity:  O(n!)
 * Space Complexity: O(n)
 * 
 * Tags: recursion, backtracking
 */

import java.util.*;

public class NQueensII {
    
    int count = 0;
    
    public int totalNQueens(int n) {
        count = 0;
        boolean[] cols = new boolean[n];
        boolean[] diag = new boolean[2 * n];
        boolean[] antiDiag = new boolean[2 * n];
        
        solve(0, n, cols, diag, antiDiag);
        return count;
    }
    
    void solve(int row, int n, boolean[] cols, boolean[] diag, boolean[] antiDiag) {
        if (row == n) {
            count++;
            return;
        }
        
        for (int col = 0; col < n; col++) {
            int d = row - col + n;
            int ad = row + col;
            
            if (cols[col] || diag[d] || antiDiag[ad]) continue;
            
            cols[col] = diag[d] = antiDiag[ad] = true;
            solve(row + 1, n, cols, diag, antiDiag);
            cols[col] = diag[d] = antiDiag[ad] = false;
        }
    }
    
    public static void main(String[] args) {
        NQueensII sol = new NQueensII();
        System.out.println("N=4: " + sol.totalNQueens(4));  // 2
        System.out.println("N=8: " + sol.totalNQueens(8));  // 92
    }
}
